text/0039-reduce-module-size.md

```diff
@@ -0,0 +1,533 @@
+that use the Lambda Layers in some network-restricted environments. Most
+alternative solutions will also have a breaking change for these customers. The
+proposed solution will have the easiest migration path for these customers.
+
+### What alternative solutions did you consider?
+
+Each category of files has different alternative solutions to reduce the size.
+
+#### .jsii Files
+
+1. The .jsii.tabl.json  and .jsii files are both “pretty-printed.” We can reduce
+   their size to 42 MB (-23%) and 30 MB (-33%) by removing all the whitespace
+   and newlines. We will not take this solution, because the space savings is
+   not as significant as compression. And, then we can maintain the benefit of
+   using pretty-printed JSON when uncompressed.
+2. Vend the Rosetta tablet file (.jsii.tabl.json) separately. On each release,
+   we would publish this file as a separate artifact on GitHub releases, and
+   reference the URI from the jsii assembly (.jsii). This solution will not be
+   pursued at this time. There are several issues with de-coupling the Rosetta
+   tablet file from the artifacts that customers install, and with requiring the
+   JSII runtime to run in an environment that has access to the public endpoint
+   where it would be hosted.
+
+#### Source Map Files (.js.map)
+
+1. Minifying the files, instead of removing them. Since the source maps are not
+   useful in their current form, it is best to just remove them. In the future,
+   we could consider adding them back in a minified form.
+2. Releasing an additional 2.X.Y-debug package with every release, that will
+   contain the source maps and source files. In addition to reducing the size of
+   the aws-cdk-lib package, we want to reduce the volume of artifacts we publish
+   to npm and other package managers in total. This would be in direct conflict
+   with that goal by doubling the amount of content that is published on each
+   release. And, there is currently no demonstrated customer need for a debug
+   package like this.
+
+#### Lambda Layer Zip Files
+
+1. Host the zip files in a static website. This would be implemented as a CDK
+   app in a new GitHub repository. The CDK CLI will download and cache these zip
+   files at synth time. The existing `lambda-layer-X` modules will be modified
+   to reference the large dependencies from a local cache. The CDK CLI will be
+   modified to update that cache at synth time, if needed. Then, the rest of
+   their usage will remain the same. The URLs for each dependency would follow a
+   predictable format, including each dependency’s version. Some automated
+   mechanism would be required to add new artifacts to the website, as new
+   versions of each dependency are released. And, to add new versions of each
+   dependency to aws-cdk-lib. The problem with this solution is that it requires
+   access to a public endpoint during synth of CDK applications. Many customers
```

Comment:
> The problem with this solution is that it requires
   access to a public endpoint during synth of CDK applications.

Doesn't the proposed chosen solution suffer from the some problem?

Reply:
True. I need a better way to phrase this. The chosen solution is slightly better because the public endpoint is one that we already know customers will have access to.